labgrid/driver/shelldriver.py

```diff
@@ -208,12 +206,27 @@ def _put_ssh_key(self, key):
                 if key['key'] == new_key['key']:
                     self.logger.info("Key already on target")
                     return
-            self.logger.info("Key not on target, mounting...")
-            self._run_check('echo "{}" > /tmp/keys'.format(keyline))
-            self._run_check('chmod 600 /tmp/keys')
-            self._run_check('mount --bind /tmp/keys ~/.ssh/authorized_keys')
-            self._run_check('chmod 700 ~/.ssh')
-            self._run_check('chmod 644 ~/.ssh/authorized_keys')
+
+        if test_write == 0 and read_keys == 0:
+            self.logger.debug("Key not on target and writeable, concatenating...")
+            self._run_check('echo "{}" >> ~/.ssh/authorized_keys'.format(keyline))
+            self._run_check("rm ~/.test")
+            return
+
+        if test_write == 0:
+            self.logger.debug("Key not on target, no ssh directory and writeable, concatenating...")
+            self._run("mkdir ~/.ssh/")
+            self._run_check("chmod 600 ~/.ssh/")
+            self._run_check('echo "{}" >> ~/.ssh/authorized_keys'.format(keyline))
+            self._run_check("rm ~/.test")
+            return
+
+        self.logger.debug("Key not on target and not writeable, using bind mount...")
+        self._run_check('mkdir -m 700 /tmp/labgrid-ssh/')
+        self._run("cat ~/.ssh/authorized_keys > /tmp/labgrid-ssh/authorized_keys")
+        self._run_check('echo "{}" >> /tmp/labgrid-ssh/authorized_keys'.format(keyline))
+        self._run_check('chmod 600 /tmp/labgrid-ssh/authorized_keys')
+        self._run_check('mount --bind /tmp/labgrid-ssh/ ~/.ssh/')
```

Comment:
Also give a useful error message if this fails (likely because ~/.ssh/ doesn't exist.

Reply:
I'm sending a warning now